src/inmanta/data/__init__.py

```diff
@@ -5507,123 +5515,254 @@ async def get_resources_for_version(
+                AND cm.released IS true
+        ), rs_with_name AS (
+            SELECT
+                rscm.*,
+                rs.name
+            FROM models AS cm
+            INNER JOIN resource_set_configuration_model AS rscm
+                ON rscm.environment = cm.environment
+                AND rscm.model = cm.version
+            INNER JOIN {ResourceSet.table_name()} AS rs
+                ON rs.environment = rscm.environment
+                AND rs.id = rscm.resource_set_id
+        -- pairwise (two-by-two) model versions to build the diff incrementally
+        ), model_pairs AS (
+            SELECT *
+            FROM (
+                SELECT
+                    first_value(version) OVER pairs AS old,
+                    last_value(version) OVER pairs AS new
+                FROM models AS cm
+                WINDOW pairs AS (ORDER BY cm.version ROWS 1 PRECEDING)
+                ORDER BY cm.version
+            )
+            -- drop first row where there is no PRECEDING for the window function
+            WHERE old != new
+        )
+        SELECT
+            reference_model.exists,
+            model_pairs.new AS version,
+            -- each row of the diff will always have at least one of the resource set ids set.
+            -- => if both are NULL, there were no records at all on the rhs of the LEFT JOIN
+            (diff.old_resource_set_id IS NULL AND diff.resource_set_id IS NULL) AS empty_model,
+            diff.name AS resource_set_name,
+            diff.resource_set_id,
+            {projection_selectors}
+        FROM reference_model
+        LEFT JOIN model_pairs ON true
+        LEFT JOIN LATERAL (
+            -- DISTINCT because join results in two null rows for shared set. ORDER BY below ensures we keep the latest one
+            SELECT DISTINCT ON (rs_new.name, rs_old.name)
+                -- if the set exists in only one of the models, report that one
+                COALESCE(rs_new.name, rs_old.name) AS name,
+                -- we're only interested in the latest id, or the absence of it
+                rs_old.resource_set_id AS old_resource_set_id,
+                rs_new.resource_set_id
+            FROM (SELECT * FROM rs_with_name WHERE model = model_pairs.old) AS rs_old
+            FULL JOIN (SELECT * FROM rs_with_name WHERE model = model_pairs.new) AS rs_new
+                ON rs_old.name = rs_new.name
+            -- filter out all sets that remained unchanged
+            WHERE rs_new.resource_set_id IS DISTINCT FROM rs_old.resource_set_id
+            -- make sure SELECT DISTINCT matches the new id for the shared set, if it exists
+            ORDER BY rs_new.name, rs_old.name, rs_new.resource_set_id NULLS LAST
+        ) AS diff
+        ON true
+        LEFT JOIN {cls.table_name()} as r
+            ON r.environment = $1
+            AND r.resource_set_id = diff.resource_set_id
+        ORDER BY model_pairs.new, diff.name, r.resource_id
```

Comment:
I find it really hard to visualise what is the output of this query

Reply:
Can you add a comment explaining what should be the output and in what format?

Reply:
I feel the same.